Dashboard totals include totals of all currencies now

feldsam opened the pull request:
Signed-off-by: Kristián Feldsam <feldsam@gmail.com>

robwent commented:
I just updated to 4.5.1 and then copied the 4 files but I now get `ERR_TOO_MANY_REDIRECTS` when trying to access the site.

I tried adding `?clear_cache=true` but still had the same issue.

robwent commented:
Yes, was working before and after I put the files back.

I didn't think composer install was needed since there's no change to the composer file.

How do I run the migrations (and how do I roll them back if it doesn't work)?

feldsam commented:
and what version you have before upgrade to 4.5.1?

`php artisan migrate`
`php artisan migrate:rollback`

robwent commented:
4.5.0
Will give it a test now

feldsam commented:
do you have debug mode enabled?

robwent commented:
Nothing to migrate.

Tried setting APP_DEBUG to true in the .env file.

In storage/logslaravel-error.log I have

    app/Http/Controllers/DashboardController.php [Line 49] => Undefined property: stdClass::$exchange_rate

feldsam commented:
try run `php artisan cache:clear`

robwent commented:
Still the same.

The error message was from earlier. If I empty out the log files and then reload the site they stay empty so that might be unrelated.

I checked the server logs and it looks like 302 redirects adding and removing /dashboard from the url.

If I try a different browser, I can get to the login screen, but then get into the loop after trying to log in.

feldsam commented:
hmm, I don't know what can cause loop

robwent commented:
It's only happening on the dashboard. I can still access other admin pages if I put the url in directly.

feldsam commented:
I can not debug if there is no error log and on my DB it is working. I suppose that you can not share me your entire DB.

robwent commented:
I'll keep looking around first and let you know if I find anything.

robwent commented:
If I uncomment this `->firstOrFail()` line then I get an error message with no redirects https://github.com/invoiceninja/invoiceninja/pull/2242/commits/5a83542d063c0c9d2c26f3500c1b5eac5eab52d2#diff-a3c24a25d054ec18e6c4956b6ea995b0R77

And if I manually set `$balancesTotals` then the dash works again. So seems to be related to items that don't have a currency set.

feldsam commented:
great, would you please paste here DB Quieries from debug console of laravel? I like to see that db query.

robwent commented:
Changing firstOrFail to just first shows:

    (1/1) ErrorException
    Trying to get property of non-object

    in DashboardController.php (line 80)
    --
    at HandleExceptions->handleError(8, 'Trying to get property of non-object', 'xxxx/app/Http/Controllers/DashboardController.php', 80, array('user' => object(User), 'viewAll' => true, 'userId' => 1, 'account' => object(Account), 'accountId' => 1, 'dashboardRepo' => object(DashboardRepository), 'metrics' => object(stdClass), 'paidToDate' => object(Collection), 'averageInvoice' => object(Collection), 'balances' => object(Collection), 'activities' =>object(Collection), 'pastDue' => object(Collection), 'upcoming' => object(Collection), 'payments' => object(Collection), 'expenses' => object(Collection), 'tasks' =>object(Collection), 'paidToDateTotal' => 37298.811264000004, 'item' => object(stdClass), 'invoiceTotal' => 42190.749993999998, 'invoiceTotalCount' => 74, 'averageInvoiceTotal' => 570.1452701891892, 'balancesTotals' => 0.0, 'currencies' => array(object(Currency), '' => null)))in DashboardController.php (line 80)

robwent commented:
DB queries:

https://gist.github.com/robwent/59d5ba762b86e9b95c1f903386285ae1

robwent commented:
Won't this always fail?

    if (! isset($currencies[$item->currency_id])) {

since currencies is initialized just before the loop as an empty array.

robwent commented:
Ignore that, I'm blind to the !

robwent commented:
`currency_id` isn't set for the second item (my default currency GBP)

feldsam commented:
looks like that when client is creating and you don't touch client currency leaving it preset to default, system internally doesn't set it and leaves it NULL in DB. So I have to fix DB query to count with NULLs and use account base currency...

feldsam commented:
try now, it should be working now

robwent commented:
After the update, I'm getting:

    (1/1) ErrorExceptionUndefined variable: account
    --
    in DashboardRepository.php (line 316)

feldsam commented:
oh, sorry, fixed now

robwent commented:
Dashboard is loading now, but it looks like I get a 500 error when clicking on the totals button:


    (1/1) ErrorExceptionUndefined property: stdClass::$invoice_text1
    --
    in Account.php (line 783)

feldsam commented:
fixed, you dont have any invoice custom fields. so I added check for that, but you still need invoice_custom field for exchange_rate.

go to /settings/products and bellow "Convert products" is Note about adding invoicce custom field.

You also need have filled exchnage rate on all invoices in foreing currency

feldsam commented:
and one more thing, there is fixer.io importer, but you need register free account and set api key in .env

robwent commented:
I have fixer set up in my env file

    EXCHANGE_RATES_ENABLED=true
    XCHANGE_RATES_URL=http://data.fixer.io/latest?access_key=MYKEY
    EXCHANGE_RATES_BASE=GBP

I'm not sure that it's working though as the currencies table shows all exchange_rate as 1.0000.
If I try and access the exchange rate url directly the api is working.

I didn't have the convert products option checked, but when I try and click on the custom fields link, I now get another error :)

    

    (2/2) ErrorExceptionUse of undefined constant ENTITY_RECURRING_QUOTE - assumed 
    'ENTITY_RECURRING_QUOTE' (View: 
    /xxxx/resources/views/accounts/invoice_settings.blade.php)
    --
    in Account.php (line 1048)

feldsam commented:
because you can not just copy files from develop to stable. I am also not sure, if fixer is fixed in stable branch. In develop it is fixed. If you want use changes from this PR, you have to use develop branch.

robwent commented:
ah ok. I'll roll my install back to the stable version, but with the last update the dash totals button is working!

feldsam commented:
thank you for testing!

feldsam commented:
This feature will be in 4.6
